lib/double-fetch-detection.c

```diff
@@ -0,0 +1,34 @@
+#include "linux/double-fetch-detection.h"
+#include <linux/context_tracking.h>
+#include <linux/entry-common.h>
+#include <linux/livepatch.h>
+#include <linux/audit.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+void addAddress(void* addr, unsigned long leng){
+           if(current->noRead==100 || current->noRead==-1)
+             return;
+           ((current->addresses)+current->noRead)->startaddress = addr;
+            ((current->addresses)+current->noRead)->len =leng;
+           current->noRead++;
+}
+
+void startSystemCall(){
+    current->addresses= (struct address*)kmalloc_array(100,sizeof(struct address),GFP_KERNEL);
```

Comment:
Hm, actually, as we're already allocating the memory dynamically, let's start with a small number of records (say, 16), and then reallocate the array if we need more.
You'll need to store the array size in `current` as well.

Reply:
okay i did that and every time it needs more memory i reallocate and make the size its double like(16,32,64..)
but can i make an upper bound for the size INT_MAX?

Reply:
Hm, interesting. Can you call dump_stack() to see in which cases this happens?

I didn't expect the number of accesses to be bigger than IOV_MAX+10 (for `readv()/writev()`, which probably have the most number of pointers passed to them), but maybe I am missing something. E.g. the kernel may be splitting the buffer passed to `write()` in fixed-size chunks (no idea if this is actually happening).

Reply:
Sorry, I misunderstood Merna's previous comment.
We've figured out in an offline chat that there's no problem with doubling the buffer size. I also suggested to set a decent upper limit (say, 1024 records) and warn if the syscall requests more (e.g. by calling `WARN_ON()`)